lib/multi.c

```diff
@@ -1155,6 +1158,26 @@ CURLMcode curl_multi_fdset(struct Curl_multi *multi,
 
 #define NUM_POLLS_ON_STACK 10
 
+#ifdef USE_WINSOCK
+static DWORD multi_getsocktype(struct Curl_multi *multi, curl_socket_t s)
+{
+  struct Curl_sh_entry *entry = NULL;
+  DWORD st = 0;
+  int sl = sizeof(st);
+
+  entry = sh_getentry(&multi->sockhash, s);
+  if(!entry)
+    return SOCK_STREAM;
+  if(entry->socketype)
+    return entry->socketype;
+  if(getsockopt(s, SOL_SOCKET, SO_TYPE, (char *)&st, &sl) == SOCKET_ERROR)
+    return SOCK_STREAM;
+
+  entry->socketype = st;
+  return st;
+}
+#endif
```

Comment:
I think the function could use a comment that it is only to be called on libcurl sockets and not the extra sockets, and also explain why it defaults to SOCK_STREAM when  no entry is available or the socket type cannot be obtained. Conversely the second check sets the type to 0 when it cannot be obtained

Reply:
I think this function *can* be called for extra sockets as they may overlap with libcurl sockets. We cannot assume that they are fully distinct or can we?

Reply:
I encourage you to test your caching method against getsockopt on 1000 sockets so we can see if it has any performance gain and is worth pursuing. For example
~~~c
  if(getsockopt(s, SOL_SOCKET, SO_TYPE, (char *)&st, &sl) != SOCKET_ERROR &&
     st == SOCK_STREAM) {
    send(s, NULL, 0, 0); /* reset FD_WRITE */
  }
~~~